### ISSUE TITLE
mosh-server deadlock on BSD ptys

### ISSUE THREAD

**Issue by cgull** [mobile-shell/mosh#692]
I discovered this a couple of days ago, and I'm a little surprised nobody's seen it before.

On FreeBSD and OS X, if you type `^S` at a program that's pushing text as fast as possible to mosh-server, mosh-server will hang:  it stops sending messages, incoming messages from mosh-client accumulate in the receive queue, and of course text output on the pty is stopped.

On FreeBSD the issue is that the pts driver effectively withdraws ready characters when it goes to a stopped state when ^S is written to it.  If mosh-server finds that both a network packet and pty input are ready, it will then receive the network packet, transmit any user input (in this case, `^S`), and only then attempt to read program output.  In this situation it blocks in that `read()`, which will never unblock because the pty is stopped...

This is arguably a bug in pty behavior, but it goes back to FreeBSD 7.2 and is probably in several versions of OS X as well, so I've coded a fix, which is just to reorder mosh's code so that it reads the pty before it tries to write anything to it.

Again, I'm really puzzled that nobody's seen this before, including me...

**Comment by keithw** [comment]
Boy! To make sure I understand: The OS's behavior is that `select()` reports that the pty fd is readable, but then when mosh-server tries to read it, `read()` blocks anyway?

That's crazy (and seems like a bug in the OS). (Why would the OS even want to withdraw **input** characters just because the user sent the stop char, described in `stty(1)` as "CHAR will stop the **output**"?)

How confident are you that reordering the event loop is good enough to fix this? Could we have the same problem if `mosh-server` wrote the stop char during some prior run of the event loop, the tty took a while to get around to processing it, `mosh-server` ran `select` and learned that the pty was readable, and **then** the tty driver processed the stop char?

**Comment by cgull** [comment]
I worded that poorly.  What happens is that `select()` gives us a read notification for the pty.  We output a `^S`, then attempt to read the pty.  That read blocks.  So the pty effectively takes back its output or our input.

I'm quite confident this fixes it.  In the FreeBSD kernel, this behavior is synchronous-- the `^S` has immediate effect, so your scenario is unlikely.  And if your scenario *were* true, *everything* that used ptys and `select()` would be having trouble.

The issue might have appeared when the `pts` driver got reimplemented a few years back.  Apple imported some FreeBSD code a few years back also, and they might have gotten that `pts` driver.  Otherwise, I can't explain the issue on OS X.

**Comment by keithw** [comment]
Hmm, how about if a program (run inside `mosh-server`) calls `tcflow( STDOUT_FILENO, TCOOFF )` at an unlucky time, i.e. between the call to `select` and the call to `read`? On Linux that triggers the same behavior as typing the stop char, but is not triggered by terminal input.

**Comment by cgull** [comment]
I don't think we can do anything about that scenario, and it would clearly be a kernel bug in that case.  In my case, it's just slightly arguable that we're responsible for the problem, because we altered the device state before reading the pty.

Well, one other thing is possible:  we could do non-blocking reads on the pty.  In order to do that, though, we'd also have to do non-blocking *writes* on the pty, too.  I think that moving mosh to a clean event-driven model where we could do this is a really good idea, but this is a bug and a fix for it.  (And a regression test.)

**Comment by keithw** [comment]
Ok, I don't object to your fix and test if that's how you want to handle it.

I do think we could get away with the nonblocking reads/writes to the pty, since afaik we already have the event-driven model (mosh-server only reads or writes from an fd after learning that it was readable/writeable from select!). So we could keep the existing code as-is, set the pty fd to nonblocking, and just hope that EAGAIN or EWOULDBLOCK are rare occurrences (perhaps logging a pedantic message).

The risk, though, would be that somehow `select` would continue reporting the fd as readable and `read` would continue giving `EWOULDBLOCK`, and then we'd get stuck in a loop.

**Comment by cgull** [comment]
Actually, mosh-server does *not* do select on writes to the pty, and mosh-client does not select on writes to its tty.  Both of them may block (it's obviously much likelier for mosh-client than mosh-server).

I also now understand why pty packet mode exists.  It fixes exactly this problem by writing a packet to the master process when termio transitions to this stopped state (or is flushed), ensuring that `read()` never blocks.  That'd be another way to fix mosh-server.

**Comment by keithw** [comment]
Ah, I stand corrected!

**Comment by cgull** [comment]
I slept on it, and I came around to your point of view on risks.  I implemented packet mode support and spent...too long working past bugs both in pty implementations and Mosh, and in the end packet mode seems to be broken too on FreeBSD.  I also didn't like my original fix and came up with a better one.

But if I'm the only one seeing this problem and it's restricted to a couple of less-common mosh-server platforms, it's not urgent to fix this.  I'd like to understand why it hasn't been seen before, and I might push a FreeBSD bug report first.

Is anybody else seeing this?
